File: delivery-frame-spec/scripts/hash_change_artifacts.py
#!/usr/bin/env python3
"""Compute artifact_revision for an OpenSpec change directory.

Hashes authoritative planning/execution files (proposal, design, tasks, specs/**,
verification when present). Ignores _test_harness and dotfiles.

Usage:
  python hash_change_artifacts.py path/to/openspec/changes/<id>
  python hash_change_artifacts.py path/to/openspec/changes/<id> --json
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path

INCLUDE_NAMES = {
    "proposal.md",
    "design.md",
    "tasks.md",
    "verification.md",
}
INCLUDE_GLOBS = ("specs/**/*.md",)


def iter_artifact_files(change_dir: Path) -> list[Path]:
    files: list[Path] = []
    for name in INCLUDE_NAMES:
        path = change_dir / name
        if path.is_file():
            files.append(path)
    for pattern in INCLUDE_GLOBS:
        files.extend(p for p in change_dir.glob(pattern) if p.is_file())
    # Stable order by relative posix path
    return sorted({p.resolve() for p in files}, key=lambda p: p.relative_to(change_dir.resolve()).as_posix())


def hash_change(change_dir: Path) -> tuple[str, list[dict[str, str]]]:
    root = change_dir.resolve()
    entries: list[dict[str, str]] = []
    h = hashlib.sha256()
    for path in iter_artifact_files(change_dir):
        rel = path.relative_to(root).as_posix()
        data = path.read_bytes()
        file_digest = hashlib.sha256(data).hexdigest()
        entries.append({"path": rel, "sha256": file_digest})
        h.update(rel.encode("utf-8"))
        h.update(b"\0")
        h.update(data)
        h.update(b"\0")
    return h.hexdigest(), entries


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("change_dir", type=Path, help="OpenSpec change directory")
    parser.add_argument("--json", action="store_true", help="Print JSON with per-file digests")
    args = parser.parse_args()
    change_dir = args.change_dir
    if not change_dir.is_dir():
        print(f"ERROR: not a directory: {change_dir}", file=sys.stderr)
        return 2
    digest, entries = hash_change(change_dir)
    if not entries:
        print("ERROR: no authoritative artifacts found", file=sys.stderr)
        return 1
    if args.json:
        print(
            json.dumps(
                {"artifact_revision": digest, "files": entries},
                ensure_ascii=False,
                indent=2,
            )
        )
    else:
        print(digest)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: delivery-frame-spec/scripts/hash_change_artifacts.mjs
#!/usr/bin/env node
// @ts-check
/**
 * Compute artifact_revision for an OpenSpec change directory.
 *
 * Hashes authoritative planning/execution files (proposal, design, tasks, specs/**,
 * verification when present). Ignores _test_harness and dotfiles.
 *
 * Usage:
 *   node hash_change_artifacts.mjs path/to/openspec/changes/<id>
 *   node hash_change_artifacts.mjs path/to/openspec/changes/<id> --json
 *
 * Zero dependencies (Node >= 18).
 */

import { createHash } from "node:crypto";
import { readFileSync, readdirSync, statSync } from "node:fs";
import { resolve, join, relative, sep } from "node:path";
import { fileURLToPath } from "node:url";
import process from "node:process";

const INCLUDE_NAMES = ["proposal.md", "design.md", "tasks.md", "verification.md"];

/** @param {string} p */
function isFile(p) {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

/**
 * Recursively collect *.md files under dir.
 * @param {string} dir
 * @returns {string[]}
 */
function mdFilesUnder(dir) {
  /** @type {string[]} */
  const out = [];
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const entry of entries) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) out.push(...mdFilesUnder(full));
    else if (entry.isFile() && entry.name.endsWith(".md")) out.push(full);
  }
  return out;
}

/** @param {string} root @param {string} p */
function relPosix(root, p) {
  return relative(root, p).split(sep).join("/");
}

/**
 * List authoritative artifact files sorted by relative posix path.
 * @param {string} changeDir
 * @returns {string[]}
 */
export function iterArtifactFiles(changeDir) {
  const root = resolve(changeDir);
  const files = new Set();
  for (const name of INCLUDE_NAMES) {
    const p = join(root, name);
    if (isFile(p)) files.add(resolve(p));
  }
  for (const p of mdFilesUnder(join(root, "specs"))) files.add(resolve(p));
  return [...files].sort((a, b) => {
    const ra = relPosix(root, a);
    const rb = relPosix(root, b);
    return ra < rb ? -1 : ra > rb ? 1 : 0;
  });
}

/**
 * Hash the change directory's authoritative artifacts.
 * Byte-for-byte compatible with the original Python implementation.
 * @param {string} changeDir
 * @returns {{ digest: string, entries: Array<{path: string, sha256: string}> }}
 */
export function hashChange(changeDir) {
  const root = resolve(changeDir);
  /** @type {Array<{path: string, sha256: string}>} */
  const entries = [];
  const h = createHash("sha256");
  for (const path of iterArtifactFiles(changeDir)) {
    const rel = relPosix(root, path);
    const data = readFileSync(path);
    entries.push({ path: rel, sha256: createHash("sha256").update(data).digest("hex") });
    h.update(Buffer.from(rel, "utf-8"));
    h.update(Buffer.from([0]));
    h.update(data);
    h.update(Buffer.from([0]));
  }
  return { digest: h.digest("hex"), entries };
}

function main() {
  const args = process.argv.slice(2);
  const wantJson = args.includes("--json");
  const positional = args.filter((a) => a !== "--json");
  if (positional.length !== 1) {
    console.error("usage: node hash_change_artifacts.mjs <change_dir> [--json]");
    return 2;
  }
  const changeDir = positional[0];
  let isDir = false;
  try {
    isDir = statSync(changeDir).isDirectory();
  } catch {
    isDir = false;
  }
  if (!isDir) {
    console.error(`ERROR: not a directory: ${changeDir}`);
    return 2;
  }
  const { digest, entries } = hashChange(changeDir);
  if (entries.length === 0) {
    console.error("ERROR: no authoritative artifacts found");
    return 1;
  }
  if (wantJson) {
    console.log(JSON.stringify({ artifact_revision: digest, files: entries }, null, 2));
  } else {
    console.log(digest);
  }
  return 0;
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  process.exit(main());
}
